torch/utils/data/datasets/decoder.py

```diff
@@ -0,0 +1,270 @@
+            elif atype == "numpy":
+                result = np.asarray(img)
+                assert result.dtype == np.uint8, "numpy image array should be type uint8, but got {}".format(result.dtype)
+                if etype == "uint8":
+                    return result
+                else:
+                    return result.astype("f") / 255.0
+            elif atype == "torch":
+                import torch
+
+                result = np.asarray(img)
+                assert result.dtype == np.uint8, "numpy image array should be type uint8, but got {}".format(result.dtype)
+
+                if etype == "uint8":
+                    result = np.array(result.transpose(2, 0, 1))
+                    return torch.tensor(result)
+                else:
+                    result = np.array(result.transpose(2, 0, 1))
+                    return torch.tensor(result) / 255.0
+            return None
+
+def imagehandler(imagespec):
+    return ImageHandler(imagespec)
+
+
+################################################################
+# torch video
+################################################################
+
+
+def torch_video(key, data):
+    extension = re.sub(r".*[.]", "", key)
+    if extension not in "mp4 ogv mjpeg avi mov h264 mpg webm wmv".split():
+        return None
+
+    # add `type: ignore` to avoid mypy's warning on import missing
+    import torchvision.io  # type: ignore
+    with tempfile.TemporaryDirectory() as dirname:
+        fname = os.path.join(dirname, f"file.{extension}")
+        with open(fname, "wb") as stream:
+            stream.write(data)
+            return torchvision.io.read_video(fname)
+
+
+################################################################
+# torchaudio
+################################################################
+
+
+def torch_audio(key, data):
+    extension = re.sub(r".*[.]", "", key)
+    if extension not in ["flac", "mp3", "sox", "wav", "m4a", "ogg", "wma"]:
+        return None
+
+    # add `type: ignore` to avoid mypy's warning on import missing
+    import torchaudio  # type: ignore
+    with tempfile.TemporaryDirectory() as dirname:
+        fname = os.path.join(dirname, f"file.{extension}")
+        with open(fname, "wb") as stream:
+            stream.write(data)
+            return torchaudio.load(fname)
+
+
+
+################################################################
+# a sample decoder
+################################################################
+
+
+class Decoder:
+    """
+    Decode key/data sets using a list of handlers.
+    For each key/data item, this iterates through the list of
+    handlers until some handler returns something other than None.
+    """
+
+    def __init__(self, handlers):
+        self.handlers = handlers
+
+    def add_decoder(self, decoder):
+        if not decoder:
+            return
+        if not self.handlers:
+            self.handlers = [decoder]
+        else:
+            self.handlers.append(decoder)
```

Comment:
It should be `add_handler` right?
`add_decoder` makes it hard to understand which one is decoder.

Reply:
I propose two APIs:
1. add_handler
2. add_decoder

Reply:
Probably adding a comments or type annotation? Anything that is callable (with 0 arg) can be passed in as a decoder.

Reply:
You mean 2 arguments, right? key and data

Reply:
Ah, yes, 2 arg.